### PR TITLE
Add a PyTorch template for custom containers

### PR DESCRIPTION
**Pull request by nnegrey**

### CODE REVIEW

**Comment by nnegrey** [comment]
Oh I should probably figure out how to set up testing too, huh?

**Comment by gogasca** [comment]
> Oh I should probably figure out how to set up testing too, huh?

Yes, if you look at the other folders, sklearn and tensorflow we have a train-local.sh script which does most of the job. 
You can do testing in a separate PR:
In GitHub should be very straightforward:
https://github.com/GoogleCloudPlatform/ai-platform-samples/tree/master/.kokoro/tests/training
https://github.com/GoogleCloudPlatform/ai-platform-samples/tree/master/.kokoro/training

File: training/pytorch/structured/custom_containers/base/README.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+* [Setup docker with Cloud Container Registry](https://cloud.google.com/container-registry/docs/pushing-and-pulling)
+* The datasets are downloaded by the Dockerfile.
+ * The Dockerfile defaults to downloading the small dataset, if you wish to modify this, you can set which files to download via the `--build-arg` flag:
+ ```
+ docker build -f Dockerfile -t $IMAGE_URI ./ --build-args train-files=GCS_PATH_TO_TRAIN_DATA eval-files=GCS_PATH_TO_EVAL_DATA
+ ```
+ * [OPTIONAL] Download the datasets using run [download-taxi.sh](../../../../datasets/download-taxi.sh) located in [datasets](../../../../datasets) folder.
+* Change the directory to this sample and run
+```
+docker build -f Dockerfile -t $IMAGE_URI ./
```
Comment (gogasca) [inline training/pytorch/structured/custom_containers/base/README.md]: Under $IMAGE_URI can we use some of the new Deep Learning containers images? as an example?

File: training/pytorch/structured/custom_containers/base/scripts/train-cloud.sh [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+echo "Training cloud ML model"
+
+# IMAGE_REPO_NAME: the image will be stored on Cloud Container Registry
+IMAGE_REPO_NAME=pytorch_taxi_container
+
+# IMAGE_TAG: an easily identifiable tag for your docker image
+IMAGE_TAG=taxi_pytorch
+
+# IMAGE_URI: the complete URI location for Cloud Container Registry
+IMAGE_URI=gcr.io/${PROJECT_ID}/${IMAGE_REPO_NAME}:${IMAGE_TAG}
```
Comment (gogasca) [inline training/pytorch/structured/custom_containers/base/scripts/train-cloud.sh]: nit: Add example using Deep Learning containers image.

File: training/pytorch/structured/custom_containers/base/trainer/experiment.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+# limitations under the License.
+
+import torch
+
+import inputs
+import model
+
+
+def train(sequential_model, train_loader, criterion, optimizer, epoch):
+    """Create the training loop"""
```
Comment (gogasca) [inline training/pytorch/structured/custom_containers/base/trainer/experiment.py]: Add method documentation:
https://github.com/google/styleguide/blob/gh-pages/pyguide.md#383-functions-and-methods

File: training/pytorch/structured/custom_containers/base/trainer/experiment.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+        # print statistics
+        running_loss += loss.item()
+        if batch_index % 10 == 9:  # print every 10 mini-batches
+            print('[epoch: %d, batch: %5d] loss: %.3f' %
+                  (epoch, batch_index + 1, running_loss / 10))
+            running_loss = 0.0
+
+
+def test(sequential_model, test_loader, criterion):
```
Comment (gogasca) [inline training/pytorch/structured/custom_containers/base/trainer/experiment.py]: Add method documentation:
https://github.com/google/styleguide/blob/gh-pages/pyguide.md#383-functions-and-methods

File: training/pytorch/structured/custom_containers/base/trainer/experiment.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+    # print statistics
+    print('\nTest set:\n\tAverage loss: {:.4f}'.format(test_loss))
+    print('\tAccuracy: {}/{} ({:.0f}%)\n'.format(
+            correct,
+            len(test_loader.sampler),
+            100. * correct / len(test_loader.sampler)))
+
+
+def run(args):
```
Comment (gogasca) [inline training/pytorch/structured/custom_containers/base/trainer/experiment.py]: same: https://github.com/google/styleguide/blob/gh-pages/pyguide.md#383-functions-and-methods

File: training/pytorch/structured/custom_containers/base/trainer/experiment.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+
+def test(sequential_model, test_loader, criterion):
+    """Test the DNN"""
+    sequential_model.eval()
+    test_loss = 0.0
+    correct = 0
+
+    with torch.no_grad():
+        for i, data in enumerate(test_loader, 0):
```
Comment (gogasca) [inline training/pytorch/structured/custom_containers/base/trainer/experiment.py]: Looks like i is not used, you can use _

File: training/pytorch/structured/custom_containers/base/trainer/model.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+def create(args):
+    """
+    Create the model, loss function, and optimizer to be used for the DNN
+    """
+    sequential_model = SequentialDNN().double()
+    criterion = nn.BCEWithLogitsLoss()
+    optimizer = optim.Adam(sequential_model.parameters(),
+                           lr=args.learning_rate,
+                           weight_decay=args.learning_rate_decay)
```
Comment (gogasca) [inline training/pytorch/structured/custom_containers/base/trainer/model.py]: would make sense to change learning_rate_decay to weight_decay?

File: training/pytorch/structured/custom_containers/base/README.md (hunk 1 above)
Comment (nnegrey) [reply training/pytorch/structured/custom_containers/base/README.md]: Looks like they aren't meant to be used for Trainer applications out of the box, but rather you can use them as a base, such as making this the Dockerfile path
```
FROM gcr.io/deeplearning-platform-release/pytorch-cpu.1-1
```

File: training/pytorch/structured/custom_containers/base/README.md [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+functionality, you can customise these parts with your own implementation.
+
+## Prerequisites
+
+* Setup your project by following the instructions in the [setup](../../../../setup/) directory.
+* [Setup docker with Cloud Container Registry](https://cloud.google.com/container-registry/docs/pushing-and-pulling)
+* The datasets are downloaded by the Dockerfile.
+ * The Dockerfile defaults to downloading the small dataset, if you wish to modify this, you can set which files to download via the `--build-arg` flag:
+ ```
+ docker build -f Dockerfile -t $IMAGE_URI ./ --build-args train-files=GCS_PATH_TO_TRAIN_DATA eval-files=GCS_PATH_TO_EVAL_DATA
```
Comment (gogasca) [inline training/pytorch/structured/custom_containers/base/README.md]: Maybe add a sample? 
gcr.io/dpe-cloud-mle/pytorch
[HOSTNAME]/[PROJECT-ID]/[IMAGE]

Comment (gogasca) [reply training/pytorch/structured/custom_containers/base/README.md]: Also add sample for Datasets "gs://" path

Comment (gogasca) [reply training/pytorch/structured/custom_containers/base/README.md]: docker build -f Dockerfile -t gcr.io/dpe-cloud-mle/pytorch ./ --build-args train-files=gs://cloud-samples-data/ml-engine/chicago_taxi/training/small/taxi_trips_train.csv eval-files=gs://cloud-samples-data/ml-engine/chicago_taxi/training/small/taxi_trips_eval.csv
unknown flag: --build-args
See 'docker build --help'.

I needed to do:

docker build -f Dockerfile -t gcr.io/dpe-cloud-mle/pytorch ./ --build-arg train-files=gs://cloud-samples-data/ml-engine/chicago_taxi/training/small/taxi_trips_train.csv --build-arg eval-files=gs://cloud-samples-data/ml-engine/chicago_taxi/training/small/taxi_trips_eval.csv

Comment (gogasca) [reply training/pytorch/structured/custom_containers/base/README.md]: Docker Version 18.06.1-ce-mac73 (26764)

File: training/pytorch/structured/custom_containers/base/README.md [hunk 9]
```diff
@@ -0,0 +1,75 @@
+# PyTorch Custom Containers Template
```
Comment (gogasca) [inline training/pytorch/structured/custom_containers/base/README.md]: Update versions below to show Pytorch instead of TensorFlow?

Comment (nnegrey) [reply training/pytorch/structured/custom_containers/base/README.md]: Whoops, updated

File: training/pytorch/structured/custom_containers/base/README.md (hunk 8 above)
Comment (nnegrey) [reply training/pytorch/structured/custom_containers/base/README.md]: Mis type. 
Updated
